### ISSUE TITLE
Microseconds are removed from timestamps

### ISSUE THREAD

**Issue by Nosfistis** [typeorm/typeorm#4964]
**Issue type:**

[ ] question
[x] bug report
[ ] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[x] `mysql` / `mariadb`
[ ] `oracle`
[ ] `postgres`
[ ] `cockroachdb`
[ ] `sqlite`
[ ] `sqljs`
[ ] `react-native`
[ ] `expo`

**TypeORM version:**

[x] `latest`
[ ] `@next`
[ ] `0.x.x` (or put your version here)

**Steps to reproduce or a small repository showing the problem:**

After #4769 timestamp columns with microsecond precision lose the microseconds, due to JS Date conversion. JS Date has a millisecond precision (like `YYYY-MM-DDTHH:mm:ss.sssZ`), therefore the extra bits are lost in the conversion at:

https://github.com/typeorm/typeorm/blob/540ce3d63ccf59a7cc0a3c7c96e0ba7fcc03ad94/src/util/DateUtils.ts#L123-L137

My current column definition is:
```Typescript
@Column({
    type: 'timestamp',
    precision: 6,
    default: () => 'CURRENT_TIMESTAMP(6)',
    update: false,
  })
  moment: string;
```
Using plain objects at inserts (not Entity objects) has the same issue. Transformers also cannot help, since the string is untouched by that time.

I suggest looking at the value having extra digits (microseconds), and avoiding conversion. I can prepare a PR.

**Comment by silentroach** [comment]
I think #4769 is not the reason of your problem. The code there is used to compare original entity properties values with modified, to update fields with changed values only. Can you reproduce this issue before this change (in 0.2.19) ?

**Comment by Nosfistis** [comment]
> 
> 
> I think #4769 is not the reason of your problem. The code there is used to compare original entity properties values with modified, to update fields with changed values only. Can you reproduce this issue before this change (in 0.2.19) ?

I did some tests and this did happen in 0.2.19 as well, no matter if the field is marked as `Date` or `string`.  It seems that this happens due to the conversions happening at the normalizer `mixedDateToDatetimeString()`, although it is not clear what happened at `timestamp` columns at 0.2.19.


> 
> 
> Also there is no microseconds in javascript `Date` as far as I know so what behavior do you expect?

Indeed `Date` is inadequate for such operations, but since most databases support microseconds this should be supported. I think the solution would be to keep the input string as-is, if microseconds are detected or append them to the normalized string.

**Comment by silentroach** [comment]
not an easy task though

**Comment by no1ysc** [comment]
Any further progress?
I have same issue on 0.2.29

**Comment by isaaccolls** [comment]
hi!
I have the same issue  😞
is there some clean workaround?

**Comment by mcholcha** [comment]
It's working for me:

"typeorm": "0.3.11"
"mysql": "2.18.1",

```
@Column({
    type: 'datetime',
    precision: 3,
    default: () => 'UTC_TIMESTAMP(3)',
})
createdAt: string;
```

It creates such a migration:
`ALTER TABLE \`some_entity\` ADD \`createdAt\` datetime(3) NOT NULL DEFAULT UTC_TIMESTAMP(3)`

**Comment by MatthiasKunnen** [comment]
In the case of Postgres, the issue is quite deep. The precision is lost in the `node-postgres` (pg) driver, see https://github.com/brianc/node-postgres/issues/1200.

It is possible to overwrite this behavior using `pg.types.setTypeParser` and preserve the original string value though `TypeOrm` will parse it as a date during hydration and loose the precision so another workaround is needed.

## Patching the Date object for fun and profit
To preserve the precision and still use the `Date` object, thereby allowing us not to change any code, the original date string as received by Postgres can be added as a property to the `Date` instance.

This is achieved using the following code:
```typescript
import * as pg from 'pg';

export const pgDateString = Symbol('pgDate');

function parsePgDate(originalParser: (value: any) => unknown) {
    return (originalValue: string) => {
        const date = originalParser(originalValue);

        if (date instanceof Date) {
            (date as any)[pgDateString] = originalValue;
        }

        return date;
    }
}

// Add all types here that should have their precision saved
const typesToOverwrite = [
    pg.types.builtins.TIMESTAMP,
    pg.types.builtins.TIMESTAMPTZ,
];

for (const typeId of typesToOverwrite) {
    pg.types.setTypeParser(typeId, parsePgDate(pg.types.getTypeParser(typeId)));
}

// Get the date with microseconds at runtime
entity.yourDate[pgDateString] -> 2022-05-25 19:44:00.034

```
Some explanation:
- While overwriting the original date parser with a simple `new Date(value)` worked in my tests, there could be edge cases. As such the original parser is preserved and executed after we are able to save the high precision value.
- We use the `pgDate` symbol to avoid naming conflicts
- Make sure to execute/import the `pg.types` code before you set up your `TypeORM` connection

**Comment by Nosfistis** [comment]
This should be easy to implement as a change (it used to work) if the current behaviour is not the intended one.

**Comment by gioboa** [comment]
I analysed this issue and the current behaviour is the intended one, JavaScript Date objects only support millisecond precision. The microseconds are immediately lost during the conversion.
Thanks 👍

**Comment by Nosfistis** [comment]
Yes indeed JS Date has that precision @gioboa, as is mentioned above. How would you suggest storing higher precision timestamps then? Should we declare timestamps as strings?

**Comment by gioboa** [comment]
Right now I think you have to use string type for it and handle the date in your business logic.

**Comment by alumni** [comment]
I think we need to spend some time to figure out a solution that would work on all drivers, there are clients that return a `Date` object even thought the timestamp precision is higher than 3.

Besides that, there's also the timezone problem with the timestamps, which also needs a solution.

One idea would be to use whatever the client returns, like a `raw` property in `ColumnOptions` that when enabled would skip the date normalization logic.

Or we could also add something like `timestampCoercion: 'string' | 'date' | 'none'` to the `DataSourceOptions` and take advantage of the DB client's connection options on some drivers.

**Comment by alumni** [comment]
Related: https://github.com/typeorm/typeorm/issues/11612

From a conversation with @gioboa:

> So `date`, `time`, `duration` and whatever data types we have related to time, imo they should always be a string
> For `timestamp`, since js has `Date`, it would be great to still work with `Date` when possible (no precision loss), unless the user wants something else
> Something similar for `longint`/`bigint`: they are strings in some clients because converting them to numbers will cause precision loss. Some clients have options to convert them to numbers if the value allows it, or even to force the conversion to a number
> Surprisingly, very few clients use JS `BigInt` (none?), probably because the current behavior is just a legacy and `BigInt` didn't catch on in JS.
